pkg/awsutils/awsutils.go

```diff
@@ -1249,55 +1252,58 @@ func (cache *EC2InstanceMetadataCache) AllocIPAddress(eniID string) error {
 	return nil
 }
 
-// GetENIIPv4Limit return IP address limit per ENI based on EC2 instance type
-func (cache *EC2InstanceMetadataCache) GetENIIPv4Limit() (int, error) {
-	eniLimits, ok := InstanceNetworkingLimits[cache.instanceType]
-	if !ok {
-		log.Errorf("Failed to get ENI IP limit due to unknown instance type %s", cache.instanceType)
-		return 0, errors.New(UnknownInstanceType)
+func (cache *EC2InstanceMetadataCache) VerifyElseFetchLimitsFromEC2() error {
+	_, ok := InstanceNetworkingLimits[cache.instanceType]
+	if ok {
+		return nil
+	}
+
+	log.Debugf("Instance type limits are missing from vpc_ip_limits.go hence making an EC2 call to fetch the limits")
+	var eniLimits InstanceTypeLimits
+	describeInstanceTypesInput := &ec2.DescribeInstanceTypesInput{InstanceTypes: []*string{aws.String(cache.instanceType)}}
+	output, err := cache.ec2SVC.DescribeInstanceTypesWithContext(context.Background(), describeInstanceTypesInput)
+	if err != nil || len(output.InstanceTypes) != 1 {
+		log.Errorf("", err)
+		return errors.New(fmt.Sprintf("Failed calling DescribeInstanceTypes for `%s`: %v", cache.instanceType, err))
+	}
+	info := output.InstanceTypes[0]
+	// Ignore any missing values
+	instanceType := aws.StringValue(info.InstanceType)
+	eniLimit := int(aws.Int64Value(info.NetworkInfo.MaximumNetworkInterfaces))
+	ipv4Limit := int(aws.Int64Value(info.NetworkInfo.Ipv4AddressesPerInterface))
+	hypervisorType := aws.StringValue(info.Hypervisor)
+	//Not checking for empty hypervisorType since have seen certain instances not getting this filled.
```

Comment:
is our code handle empty hypervisorType properly during usage?

Reply:
Yes, there are certain instances where we are not getting hypervisorType - https://github.com/aws/amazon-vpc-cni-k8s/blob/8cb5dc2815f05818303b45ff8a55a68df890e6b7/pkg/awsutils/vpc_ip_resource_limit.go#L26. So I am checking not != nitro - https://github.com/aws/amazon-vpc-cni-k8s/blob/8cb5dc2815f05818303b45ff8a55a68df890e6b7/pkg/ipamd/ipamd.go#L370